packages/slate/src/models/schema.js

```diff
@@ -126,24 +127,33 @@ class Schema extends Record(DEFAULTS) {
   }
 
   /**
-   * Return the decorators for an `object`.
+   * Return the decorations for an `object`.
    *
    * This method is private, because it should always be called on one of the
    * often-changing immutable objects instead, since it will be memoized for
    * much better performance.
    *
    * @param {Mixed} object
-   * @return {Array}
+   * @return {List<Selection>}
    */
 
-  __getDecorators(object) {
-    return this.rules
-      .filter(rule => rule.decorate && rule.match(object))
-      .map((rule) => {
-        return (text) => {
-          return rule.decorate(text, object)
-        }
+  __getDecorations(object) {
+    const array = []
+
+    this.rules.forEach((rule) => {
+      if (!rule.decorate) return
+      if (!rule.match(object)) return
+
+      const decorations = rule.decorate(object)
```

Comment:
So the `decorate` calls themselves now live under a memoized function, which is nice :)